### ISSUE TITLE
[Documentation] How to authenticate OAuth with PKCE ?

### ISSUE THREAD

**Issue by raboon** [AzureAD/microsoft-authentication-library-for-dotnet#2043]
### Please check those that apply

- documentation doesn't exist
- documentation needs clarification
- needs example

### Description Of The Issue

Is it possible to provide an example of how to authenticate with OAuth + PKCE ?
My authentication and authorization server is not Azure.

**Comment by jmprieur** [comment]
@raboon 
- Do you mean in public client applications? (desktop/mobile)? In which case nothing needs to be done, as MSAL.NET already uses OAuth code flow + PKCE for all the public client application interactive token acquisition;
- Or do you mean in confidential client applications ? (that is in this case Web apps) ? For this, MSAL.NET does not support it yet (and to be fair confidential client applications have other ways of securing their identity using the client credentials). There is a feature request for this: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1473, and it will be processed by Microsoft.Identity.Web (https://github.com/AzureAD/microsoft-identity-web/issues/470)

**Comment by raboon** [comment]
I want to build it as part of the outlook add-in.

I tried with public client application. I tried from Console example in **MsalNetExt**.
i set 
`var pca = CreatePublicClient(url);
private static IPublicClientApplication CreatePublicClient(string authority)
        {
            var appBuilder = PublicClientApplicationBuilder.Create(Config.ClientId) 
                .WithAuthority(authority)
                .WithRedirectUri("https://localhost:4200");  
            var app = appBuilder.Build();
            Console.WriteLine($"Built public client");
            return app;
        }`

And I received as below
`
 1. Acquire Token using Username and Password - for TEST only, do not use in production!
                        2. Acquire Token using Device Code Flow
                        3. Acquire Token Interactive
                        4. Acquire Token Silent
                        5. Display Accounts (reads the cache)
                        6. Acquire Token U/P and Silent in a loop
                        7. Write to file with delay
                        c. Clear cache
                        e. Expire Access Tokens (TEST only!)
                        x. Exit app
                    Enter your Selection:
3
Exception : MSAL.NetCore.4.18.0.0.MsalServiceException:
        ErrorCode: invalid_instance
Microsoft.Identity.Client.MsalServiceException: AADSTS50049: Unknown or invalid instance.
Trace ID: e79c5ff3-9f0c-48fb-aade-dc4982b78000
Correlation ID: 53afeaa4-9c4a-401c-98c0-ad935ff833e2
Timestamp: 2020-09-16 10:34:48Z
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.ThrowServerException(HttpResponse response, RequestContext requestContext)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.CreateResponse[T](HttpResponse response, RequestContext requestContext)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.ExecuteRequestAsync[T](Uri endPoint, HttpMethod method, RequestContext requestContext, Boolean expectErrorsOn200OK, Boolean addCommonHeaders)
   at Microsoft.Identity.Client.OAuth2.OAuth2Client.DiscoverAadInstanceAsync(Uri endPoint, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.NetworkMetadataProvider.SendInstanceDiscoveryRequestAsync(Uri authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.NetworkMetadataProvider.FetchAllDiscoveryMetadataAsync(Uri authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.NetworkMetadataProvider.GetMetadataAsync(Uri authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.Discovery.InstanceDiscoveryManager.FetchNetworkMetadataOrFallbackAsync(RequestContext requestContext, Uri authorityUri, Boolean autoDetectRegion)
   at Microsoft.Identity.Client.Instance.Discovery.InstanceDiscoveryManager.GetMetadataEntryAsync(String authority, RequestContext requestContext)
   at Microsoft.Identity.Client.Instance.AuthorityEndpoints.UpdateAuthorityWithPreferredNetworkHostAsync(AuthenticationRequestParameters requestParameters)
   at Microsoft.Identity.Client.Instance.AuthorityEndpoints.UpdateAuthorityEndpointsAsync(AuthenticationRequestParameters requestParameters)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.ResolveAuthorityEndpointsAsync()
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenInteractiveParameters interactiveParameters, CancellationToken cancellationToken)
   at ManualTestApp.Program.Main(String[] args) in C:\Users\mkawsar\Downloads\microsoft-authentication-extensions-for-dotnet-master\sample\ManualTestApp\Program.cs:line 97
        StatusCode: 400
        ResponseBody: {"error":"invalid_instance","error_description":"AADSTS50049: Unknown or invalid instance.\r\nTrace ID: e79c5ff3-9f0c-48fb-aade-dc4982b78000\r\nCorrelation ID: 53afeaa4-9c4a-401c-98c0-ad935ff833e2\r\nTimestamp: 2020-09-16 10:34:48Z","error_codes":[50049],"timestamp":"2020-09-16 10:34:48Z","trace_id":"e79c5ff3-9f0c-48fb-aade-dc4982b78000","correlation_id":"53afeaa4-9c4a-401c-98c0-ad935ff833e2","error_uri":"https://login.microsoftonline.com/error?code=50049"}
        Headers: Cache-Control: max-age=86400, private
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, OPTIONS
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
client-request-id: 53afeaa4-9c4a-401c-98c0-ad935ff833e2
x-ms-request-id: e79c5ff3-9f0c-48fb-aade-dc4982b78000
x-ms-ests-server: 2.1.11021.16 - DUB2 ProdSlices
Set-Cookie: fpc=AnCrNxUex9tMnKWAJqxiJ90; expires=Fri, 16-Oct-2020 10:34:48 GMT; path=/; secure; HttpOnly; SameSite=None, x-ms-gateway-slice=prod; path=/; secure; httponly, stsservicecookie=ests; path=/; secure; samesite=none; httponly
Date: Wed, 16 Sep 2020 10:34:47 GMT`
